### PR TITLE
[WIP] experiment with a v2 serialization format

### PR DESCRIPTION
**Pull request by lavalamp**
```
BenchmarkFieldSet/serialize-size-20-12             20000             85121 ns/op          245760 B/op          1 allocs/op
BenchmarkFieldSet/serialize-20-12                 100000             16411 ns/op            2073 B/op         23 allocs/op
BenchmarkFieldSet/deserialize-20-12                50000             39908 ns/op           10965 B/op        215 allocs/op
BenchmarkFieldSet/serializeV2-size-20-12           20000             77776 ns/op          229376 B/op          1 allocs/op
BenchmarkFieldSet/serializeV2-20-12               200000             11355 ns/op            1766 B/op         54 allocs/op
BenchmarkFieldSet/deserializeV2-20-12              50000             26733 ns/op            9819 B/op        139 allocs/op
BenchmarkFieldSet/serialize-size-50-12              5000            238835 ns/op          696320 B/op          1 allocs/op
BenchmarkFieldSet/serialize-50-12                  50000             37537 ns/op            5282 B/op         48 allocs/op
BenchmarkFieldSet/deserialize-50-12                20000             84362 ns/op           23456 B/op        618 allocs/op
BenchmarkFieldSet/serializeV2-size-50-12           10000            219826 ns/op          638976 B/op          1 allocs/op
BenchmarkFieldSet/serializeV2-50-12                50000             27731 ns/op            6188 B/op        146 allocs/op
BenchmarkFieldSet/deserializeV2-50-12              30000             58751 ns/op           20203 B/op        400 allocs/op
BenchmarkFieldSet/serialize-size-100-12              500           2853465 ns/op         2310144 B/op          1 allocs/op
BenchmarkFieldSet/serialize-100-12                 10000            114115 ns/op           16990 B/op         93 allocs/op
BenchmarkFieldSet/deserialize-100-12                5000            237687 ns/op           70740 B/op       2163 allocs/op
BenchmarkFieldSet/serializeV2-size-100-12            500           2613701 ns/op         2220032 B/op          1 allocs/op
BenchmarkFieldSet/serializeV2-100-12               20000             87290 ns/op           17660 B/op        491 allocs/op
BenchmarkFieldSet/deserializeV2-100-12             10000            156089 ns/op           59673 B/op       1424 allocs/op
BenchmarkFieldSet/serialize-size-500-12              100          15700207 ns/op        11509760 B/op          1 allocs/op
BenchmarkFieldSet/serialize-500-12                  3000            598173 ns/op           77419 B/op        470 allocs/op
BenchmarkFieldSet/deserialize-500-12                2000           1163462 ns/op          328855 B/op      10781 allocs/op
BenchmarkFieldSet/serializeV2-size-500-12           1000           2266963 ns/op        11059200 B/op          1 allocs/op
BenchmarkFieldSet/serializeV2-500-12                3000            504621 ns/op          108949 B/op       2460 allocs/op
BenchmarkFieldSet/deserializeV2-500-12              2000            721080 ns/op          273557 B/op       7085 allocs/op
BenchmarkFieldSet/serialize-size-1000-12             100          30542226 ns/op        24870912 B/op          1 allocs/op
BenchmarkFieldSet/serialize-1000-12                 1000           1132630 ns/op          173221 B/op        999 allocs/op
BenchmarkFieldSet/deserialize-1000-12               1000           2043855 ns/op          721161 B/op      23545 allocs/op
BenchmarkFieldSet/serializeV2-size-1000-12           200           5231563 ns/op        24002560 B/op          1 allocs/op
BenchmarkFieldSet/serializeV2-1000-12               2000           1180381 ns/op          222182 B/op       5393 allocs/op
BenchmarkFieldSet/deserializeV2-1000-12             1000           1491113 ns/op          601350 B/op      15550 allocs/op

```

### CODE REVIEW

**Comment by lavalamp** [comment]
This is 4% smaller than v1, significantly faster to read, and significantly slower to write (but that can almost certainly be fixed).

Cons: ugly

**Comment by lavalamp** [comment]
In trying to figure out where these allocs are coming from I've made the v1 serialize do 4x fewer allocs...

**Comment by lavalamp** [comment]
Effect on existing benchmarks:

```
benchmark                                  old ns/op     new ns/op     delta
BenchmarkFieldSet/serialize-20-12          15352         15551         +1.30%
BenchmarkFieldSet/serialize-50-12          37788         33236         -12.05%
BenchmarkFieldSet/serialize-100-12         112891        96619         -14.41%
BenchmarkFieldSet/serialize-500-12         605046        502300        -16.98%
BenchmarkFieldSet/serialize-1000-12        1291668       997306        -22.79%

benchmark                                  old allocs     new allocs     delta
BenchmarkFieldSet/serialize-20-12          23             4              -82.61%
BenchmarkFieldSet/serialize-50-12          48             5              -89.58%
BenchmarkFieldSet/serialize-100-12         94             5              -94.68%
BenchmarkFieldSet/serialize-500-12         470            34             -92.77%
BenchmarkFieldSet/serialize-1000-12        998            213            -78.66%
```


Final perf numbers:

```
BenchmarkFieldSet/serialize-size-20-12     20000             83303 ns/op          245760 B/op          1 allocs/op
BenchmarkFieldSet/serialize-20-12         100000             15551 ns/op            3178 B/op          4 allocs/op
BenchmarkFieldSet/deserialize-20-12        50000             37749 ns/op           10959 B/op        215 allocs/op
BenchmarkFieldSet/serializeV2-size-20-12                   20000             80997 ns/op          229376 B/op          1 allocs/op
BenchmarkFieldSet/serializeV2-20-12                       200000              9205 ns/op            3577 B/op          5 allocs/op
BenchmarkFieldSet/deserializeV2-20-12                      50000             24750 ns/op            9818 B/op        139 allocs/op
BenchmarkFieldSet/serialize-size-50-12                      5000            222826 ns/op          696320 B/op          1 allocs/op
BenchmarkFieldSet/serialize-50-12                          50000             33236 ns/op            6990 B/op          5 allocs/op
BenchmarkFieldSet/deserialize-50-12                        20000             80038 ns/op           23436 B/op        617 allocs/op
BenchmarkFieldSet/serializeV2-size-50-12                   10000            206999 ns/op          638976 B/op          1 allocs/op
BenchmarkFieldSet/serializeV2-50-12                       100000             22421 ns/op           10298 B/op          7 allocs/op
BenchmarkFieldSet/deserializeV2-50-12                      30000             55440 ns/op           20205 B/op        401 allocs/op
BenchmarkFieldSet/serialize-size-100-12                      500           2778816 ns/op         2301952 B/op          1 allocs/op
BenchmarkFieldSet/serialize-100-12                         20000             96619 ns/op           17437 B/op          5 allocs/op
BenchmarkFieldSet/deserialize-100-12                       10000            236002 ns/op           70630 B/op       2160 allocs/op
BenchmarkFieldSet/serializeV2-size-100-12                   5000            409521 ns/op         2211840 B/op          1 allocs/op
BenchmarkFieldSet/serializeV2-100-12                       20000             66361 ns/op           34172 B/op         12 allocs/op
BenchmarkFieldSet/deserializeV2-100-12                     10000            154700 ns/op           59526 B/op       1421 allocs/op
BenchmarkFieldSet/serialize-size-500-12                      100          15555218 ns/op        11517952 B/op          1 allocs/op
BenchmarkFieldSet/serialize-500-12                          3000            502300 ns/op           57654 B/op         34 allocs/op
BenchmarkFieldSet/deserialize-500-12                        2000           1167272 ns/op          329298 B/op      10795 allocs/op
BenchmarkFieldSet/serializeV2-size-500-12                    100          15514742 ns/op        11067392 B/op          1 allocs/op
BenchmarkFieldSet/serializeV2-500-12                        5000            379041 ns/op          189197 B/op         32 allocs/op
BenchmarkFieldSet/deserializeV2-500-12                      2000            755009 ns/op          273737 B/op       7093 allocs/op
BenchmarkFieldSet/serialize-size-1000-12                     100          27864919 ns/op        24846336 B/op          1 allocs/op
BenchmarkFieldSet/serialize-1000-12                         2000            997306 ns/op          128769 B/op        213 allocs/op
BenchmarkFieldSet/deserialize-1000-12                       1000           2079805 ns/op          720717 B/op      23530 allocs/op
BenchmarkFieldSet/serializeV2-size-1000-12                   100          13422513 ns/op        23969792 B/op          1 allocs/op
BenchmarkFieldSet/serializeV2-1000-12                       2000           1002787 ns/op          420510 B/op         58 allocs/op
BenchmarkFieldSet/deserializeV2-1000-12                     1000           1769985 ns/op          600941 B/op      15540 allocs/op
```

A possible improvement is to replace the two character code with a single number.

**Comment by lavalamp** [comment]
TL;DR:
* deserialize 15% faster, 34% fewer allocs
* serialize about even (accidentally sped up existing code, so 20% speed / 95% alloc improvement)
* output 4% smaller, could be improved slightly still
* probably the ugliest output format I've ever made

**Comment by lavalamp** [comment]
I did my optimization-- it was much bigger than expected, output size is now 25% smaller.

My test uses a bunch of short field names which probably exaggerates the gain.

**Comment by lavalamp** [comment]
```
BenchmarkFieldSet/insert-20-12            100000             18564 ns/op            4750 B/op         78 allocs/op
BenchmarkFieldSet/has-20-12              2000000               748 ns/op              96 B/op          2 allocs/op
BenchmarkFieldSet/serialize-size-20-12     20000             83460 ns/op          245760 B/op          1 allocs/op
BenchmarkFieldSet/serialize-20-12         100000             15461 ns/op            3178 B/op          4 allocs/op
BenchmarkFieldSet/deserialize-20-12        50000             37291 ns/op           10958 B/op        215 allocs/op
BenchmarkFieldSet/serializeV2-size-20-12                   20000             61313 ns/op          180224 B/op          1 allocs/op
BenchmarkFieldSet/serializeV2-20-12                       200000              9288 ns/op            2907 B/op          5 allocs/op
BenchmarkFieldSet/deserializeV2-20-12                      50000             24879 ns/op            9818 B/op        139 allocs/op
BenchmarkFieldSet/union-20-12                             200000             12019 ns/op            4686 B/op         53 allocs/op
BenchmarkFieldSet/intersection-20-12                      200000              5223 ns/op            1115 B/op         29 allocs/op
BenchmarkFieldSet/difference-20-12                        200000              8880 ns/op            2743 B/op         42 allocs/op
BenchmarkFieldSet/insert-50-12                             30000             59676 ns/op           16259 B/op        246 allocs/op
BenchmarkFieldSet/has-50-12                              2000000               843 ns/op             160 B/op          2 allocs/op
BenchmarkFieldSet/serialize-size-50-12                      5000            218392 ns/op          696321 B/op          1 allocs/op
BenchmarkFieldSet/serialize-50-12                          50000             35576 ns/op            6990 B/op          5 allocs/op
BenchmarkFieldSet/deserialize-50-12                        20000             85205 ns/op           23436 B/op        617 allocs/op
BenchmarkFieldSet/serializeV2-size-50-12                   10000            163342 ns/op          499712 B/op          1 allocs/op
BenchmarkFieldSet/serializeV2-50-12                       100000             22821 ns/op            7771 B/op          6 allocs/op
BenchmarkFieldSet/deserializeV2-50-12                      30000             57286 ns/op           20205 B/op        401 allocs/op
BenchmarkFieldSet/union-50-12                              50000             29596 ns/op           10382 B/op        133 allocs/op
BenchmarkFieldSet/intersection-50-12                      200000             10269 ns/op            2352 B/op         61 allocs/op
BenchmarkFieldSet/difference-50-12                        100000             21140 ns/op            6788 B/op        100 allocs/op
BenchmarkFieldSet/insert-100-12                            10000            158620 ns/op           60117 B/op        813 allocs/op
BenchmarkFieldSet/has-100-12                             1000000              1142 ns/op             309 B/op          3 allocs/op
BenchmarkFieldSet/serialize-size-100-12                      500           2598228 ns/op         2301952 B/op          1 allocs/op
BenchmarkFieldSet/serialize-100-12                         20000             96142 ns/op           17437 B/op          5 allocs/op
BenchmarkFieldSet/deserialize-100-12                       10000            236616 ns/op           70631 B/op       2160 allocs/op
BenchmarkFieldSet/serializeV2-size-100-12                   2000            786500 ns/op         1753089 B/op          1 allocs/op
BenchmarkFieldSet/serializeV2-100-12                       20000             68766 ns/op           29106 B/op         11 allocs/op
BenchmarkFieldSet/deserializeV2-100-12                     10000            153835 ns/op           59522 B/op       1421 allocs/op
BenchmarkFieldSet/union-100-12                             20000             57937 ns/op           25845 B/op        257 allocs/op
BenchmarkFieldSet/intersection-100-12                     100000             20342 ns/op            4672 B/op        124 allocs/op
BenchmarkFieldSet/difference-100-12                        30000             47605 ns/op           16167 B/op        214 allocs/op
BenchmarkFieldSet/insert-500-12                             2000            779319 ns/op          315024 B/op       4141 allocs/op
BenchmarkFieldSet/has-500-12                             1000000              1416 ns/op             351 B/op          3 allocs/op
BenchmarkFieldSet/serialize-size-500-12                      100          15939794 ns/op        11517952 B/op          1 allocs/op
BenchmarkFieldSet/serialize-500-12                          3000            498659 ns/op           57657 B/op         35 allocs/op
BenchmarkFieldSet/deserialize-500-12                        2000           1025577 ns/op          329299 B/op      10795 allocs/op
BenchmarkFieldSet/serializeV2-size-500-12                   1000           1237233 ns/op         8757248 B/op          1 allocs/op
BenchmarkFieldSet/serializeV2-500-12                        5000            399111 ns/op          134459 B/op         26 allocs/op
BenchmarkFieldSet/deserializeV2-500-12                      2000            741799 ns/op          273732 B/op       7092 allocs/op
BenchmarkFieldSet/union-500-12                              5000            297643 ns/op          123928 B/op       1467 allocs/op
BenchmarkFieldSet/intersection-500-12                      10000            125919 ns/op           27854 B/op        740 allocs/op
BenchmarkFieldSet/difference-500-12                        10000            276029 ns/op           76096 B/op       1161 allocs/op
BenchmarkFieldSet/insert-1000-12                            1000           1557918 ns/op          692177 B/op       8979 allocs/op
BenchmarkFieldSet/has-1000-12                            1000000              1531 ns/op             377 B/op          3 allocs/op
BenchmarkFieldSet/serialize-size-1000-12                     100          19310013 ns/op        24854528 B/op          1 allocs/op
BenchmarkFieldSet/serialize-1000-12                         2000           1144653 ns/op          128569 B/op        207 allocs/op
BenchmarkFieldSet/deserialize-1000-12                       1000           2154122 ns/op          721228 B/op      23546 allocs/op
BenchmarkFieldSet/serializeV2-size-1000-12                   100          20085243 ns/op        18997248 B/op          1 allocs/op
BenchmarkFieldSet/serializeV2-1000-12                       2000            703047 ns/op          297354 B/op         47 allocs/op
BenchmarkFieldSet/deserializeV2-1000-12                     1000           1553104 ns/op          600667 B/op      15529 allocs/op
BenchmarkFieldSet/union-1000-12                             3000            550893 ns/op          239283 B/op       2863 allocs/op
BenchmarkFieldSet/intersection-1000-12                      5000            239522 ns/op           50248 B/op       1332 allocs/op
BenchmarkFieldSet/difference-1000-12                        3000            534929 ns/op          157324 B/op       2241 allocs/op
```

OK, this now is just plain 20-30% better than the first format in all dimensions.

Take a look at the golden data test and see if you can stomach it.

This needs a bit of clean up before it gets submitted.
